numpy/_typing/_array_like.py

```diff
@@ -148,6 +149,7 @@ def __array_function__(
     dtype[bytes_],
     bytes,
 ]
+_ArrayLikeString_co: TypeAlias = _SupportsArray[StringDType]
```

Comment:
I believe the `_co` suffix refers to it being "coerceable to [the relevant dtype]". 
So in this case, `_co` suffix might be misleading.

Removing it would probably be the easiest option.

Otherwise, (nested sequences of) `builtins.str` and `numpy.str_` array-likes should also be included, because

```pycon
>>> np.array(np.array(["True", "False", "FileNotFound"], np.str_), dtype=np.dtypes.StringDType())
array(['True', 'False', 'FileNotFound'], dtype=StringDType())
```

You could also consider doing both,  i.e. remove the `_co` from the current alias, and add a new `_co` one, that also includes these `str` and `str_` array-likes.

This was, you could type the first param of (the last overload of)  `numpy.strings.index` as a `_ArrayLikeString`, and the second param as `_ArrayLikeString_co`.

Reply:
The behavior you saw with `index` is a bug I think (probably a missing promoter). 

That said, if I want to use `_ArrayLikeString_co` and have it mean that *all* strings are coercible to StringDType, suddenly that becomes context dependent. One of the arguments has to be explicitly a StringDType array for the promoters to fire, otherwise NumPy would use the unicode DType loops. So, non-StringDType arguments *are* coercible, but not in the general case. Is there a way to express that in the Python type system?